draft-ietf-quic-transport.md

```diff
@@ -2884,32 +2929,39 @@ frame that signals the error.  Where this specification identifies error
 conditions, it also identifies the error code that is used.
 
 A stateless reset ({{stateless-reset}}) is not suitable for any error that can
-be signaled with a CONNECTION_CLOSE or RST_STREAM frame.  A stateless reset MUST
-NOT be used by an endpoint that has the state necessary to send a frame on the
-connection.
+be signaled with a TRANSPORT_CLOSE, APPLICATION_CLOSE, or RST_STREAM frame.  A
+stateless reset MUST NOT be used by an endpoint that has the state necessary to
+send a frame on the connection.
 
 
 ## Connection Errors
 
 Errors that result in the connection being unusable, such as an obvious
 violation of protocol semantics or corruption of state that affects an entire
-connection, MUST be signaled using a CONNECTION_CLOSE frame
-({{frame-connection-close}}). An endpoint MAY close the connection in this
-manner, even if the error only affects a single stream.
+connection, MUST be signaled using a TRANSPORT_CLOSE or APPLICATION_CLOSE frame
+({{frame-transport-close}}, {{frame-application-close}}). An endpoint MAY close
+the connection in this manner, even if the error only affects a single stream.
 
-A CONNECTION_CLOSE frame could be sent in a packet that is lost.  An endpoint
-SHOULD be prepared to retransmit a packet containing a CONNECTION_CLOSE frame if
-it receives more packets on a terminated connection.  Limiting the number of
-retransmissions and the time over which this final packet is sent limits the
-effort expended on terminated connections.
+Application protocols can signal application-specific protocol errors using the
+APPLICATION_CLOSE frame.  Errors that are specific to the transport, including
+all those described in this document, are carried in a TRANSPORT_CLOSE frame.
+Other than the type of error code they carry, these frames are identical in
+format and semantics.
 
-An endpoint that chooses not to retransmit packets containing CONNECTION_CLOSE
-risks a peer missing the first such packet.  The only mechanism available to an
-endpoint that continues to receive data for a terminated connection is to use
-the stateless reset process ({{stateless-reset}}).
+A TRANSPORT_CLOSE or APPLICATION_CLOSE frame could be sent in a packet that is
+lost.  An endpoint SHOULD be prepared to retransmit a packet containing either
+frame type if it receives more packets on a terminated connection.  Limiting the
+number of retransmissions and the time over which this final packet is sent
+limits the effort expended on terminated connections.
 
-An endpoint that receives an invalid CONNECTION_CLOSE frame MUST NOT signal the
-existence of the error to its peer.
+An endpoint that chooses not to retransmit packets containing TRANSPORT_CLOSE or
+APPLICATION_CLOSE risks a peer missing the first such packet.  The only
+mechanism available to an endpoint that continues to receive data for a
+terminated connection is to use the stateless reset process
+({{stateless-reset}}).
```

Comment:
editorial nit: I would merge this para with the one above, since this one follows from the earlier.